optaplanner-core/src/test/java/org/optaplanner/core/config/solver/EnvironmentModeTest.java

```diff
@@ -0,0 +1,237 @@
+            case FAST_ASSERT:
+                assertIllegalStateExceptionWhileSolving("Score corruption analysis could not be generated ");
+                break;
+            case REPRODUCIBLE:
+            case NON_REPRODUCIBLE:
+            case PRODUCTION:
+                // no exception expected
+                break;
+            default:
+                throw new AssertionError("Missing case for " + environmentMode);
+        }
+    }
+
+    private void assertReproducibility(Solver solver1, Solver solver2) {
+        assertThat(areGeneratingSameNumbers(
+                ((DefaultSolver) solver1).getRandomFactory(),
+                ((DefaultSolver) solver2).getRandomFactory())
+        )
+                .as("Random factories generate different random values. This is 100% wrong.")
+                .isTrue();
+        assertThat(areScoreSeriesTheSame(solver1, solver2))
+                .as("Score steps are different and should be the same!")
+                .isTrue();
+    }
+
+    private void assertNonReproducibility(Solver solver1, Solver solver2) {
+        assertThat(areGeneratingSameNumbers(
+                ((DefaultSolver) solver1).getRandomFactory(),
+                ((DefaultSolver) solver2).getRandomFactory())
+        )
+                .as("Random factories generate exactly same results. "
+                            + "It can happen but the probability is very low. Run test again")
+                .isFalse();
+        assertThat(areScoreSeriesTheSame(solver1, solver2))
+                .as("Score steps are same and should be different! "
+                            + "This might be possible because searchSpace is not infinite and "
+                            + "two different random scenarios can have same results. Run test again.")
+                .isFalse();
+    }
+
+    private void assertIllegalStateExceptionWhileSolving(String exceptionMessage) {
+        assertThatExceptionOfType(IllegalStateException.class)
+                .isThrownBy(() -> PlannerTestUtils.solve(solverConfig, inputProblem))
+                .withMessageContaining(exceptionMessage);
+    }
+
+    private boolean areScoreSeriesTheSame(Solver<TestdataSolution> solver, Solver<TestdataSolution> solver2) {
+        boolean areSame = true;
+
+        TestdataStepScoreListener listener = new TestdataStepScoreListener();
+        TestdataStepScoreListener listener2 = new TestdataStepScoreListener();
+
+        ((DefaultSolver<TestdataSolution>) solver).addPhaseLifecycleListener(listener);
+        ((DefaultSolver<TestdataSolution>) solver2).addPhaseLifecycleListener(listener2);
+
+        for (int i = 0; i < NUMBER_OF_TIMES_RUN && areSame; i++) {
+            solver.solve(inputProblem);
+            solver2.solve(inputProblem);
+
+            List<Integer> scoreTimeSeries = listener.getScores();
+            List<Integer> scoreTimeSeries2 = listener2.getScores();
+
+            areSame = scoreTimeSeries.equals(scoreTimeSeries2);
```

Comment:
Personally, I'd add the `Assertion` here. This way, the failure message will be more descriptive than ("false was not true").

Reply:
Since you want to check multiple scores in a loop, you may want to give `SoftAssertions` a try.